runtime/src/main.c

```diff
@@ -380,8 +454,37 @@ main(int argc, char **argv)
 #ifdef LOG_MODULE_LOADING
 	debuglog("Parsing modules file [%s]\n", argv[1]);
 #endif
-	if (module_alloc_from_json(argv[1])) panic("failed to initialize module(s) defined in %s\n", argv[1]);
+	const char *json_path    = argv[1];
+	char       *json_buf     = NULL;
+	ssize_t     json_buf_len = load_file_into_buffer(json_path, &json_buf);
+	if (unlikely(json_buf_len <= 0)) panic("failed to initialize module(s) defined in %s\n", json_path);
+
+	struct module_config *module_config_vec;
+
+	int module_config_vec_len = parse_json(json_buf, json_buf_len, &module_config_vec);
+	if (module_config_vec_len < 0) { exit(-1); }
+	free(json_buf);
 
+	for (int module_idx = 0; module_idx < module_config_vec_len; module_idx++) {
+		/* Automatically calls listen */
+		struct module *module = module_alloc(&module_config_vec[module_idx]);
+		if (unlikely(module == NULL)) panic("failed to initialize module(s) defined in %s\n", json_path);
+
+		int rc = module_database_add(module);
+		if (rc < 0) {
+			panic("Module database full!\n");
+			exit(-1);
+		}
+
+		/* Start listening for requests */
+		rc = module_listen(module);
+		if (rc < 0) exit(-1);
+	}
+
+	for (int module_idx = 0; module_idx < module_config_vec_len; module_idx++) {
+		module_config_deinit(&module_config_vec[module_idx]);
```

Comment:
`free(&module_config_vec[module_idx])` after deinit?

Reply:
It's an array of structures, not pointers. See `json.h:184`

```c
*module_config_vec = (struct module_config *)calloc(module_config_vec_len, sizeof(struct module_config));
```